lib/class-wp-rest-search-controller.php

```diff
@@ -0,0 +1,832 @@
+					'protected' => (bool) $post->post_password,
+				);
+			} else {
+				$data['content'] = array(
+					'raw'       => '',
+					'rendered'  => '',
+					'protected' => false,
+				);
+			}
+		}
+
+		if ( ! empty( $schema['properties']['excerpt'] ) ) {
+			if ( post_type_supports( $post->post_type, 'excerpt' ) ) {
+				/** This filter is documented in wp-includes/post-template.php */
+				$excerpt         = apply_filters( 'the_excerpt', apply_filters( 'get_the_excerpt', $post->post_excerpt, $post ) );
+				$data['excerpt'] = array(
+					'raw'       => $post->post_excerpt,
+					'rendered'  => post_password_required( $post ) ? '' : $excerpt,
+					'protected' => (bool) $post->post_password,
+				);
+			} else {
+				$data['excerpt'] = array(
+					'raw'       => '',
+					'rendered'  => '',
+					'protected' => false,
+				);
+			}
+		}
+
+		if ( ! empty( $schema['properties']['author'] ) ) {
+			if ( post_type_supports( $post->post_type, 'author' ) ) {
+				$data['author'] = (int) $post->post_author;
+			} else {
+				$data['author'] = 0;
+			}
+		}
+
+		$context = ! empty( $request['context'] ) ? $request['context'] : 'view';
+		$data    = $this->add_additional_fields_to_object( $data, $request );
+		$data    = $this->filter_response_by_context( $data, $context );
+
+		// Wrap the data in a response object.
+		$response = rest_ensure_response( $data );
+
+		$response->add_links( $this->prepare_links( $post ) );
+
+		return $response;
+	}
+
+	/**
+	 * Checks the post_date_gmt or modified_gmt and prepare any post or
+	 * modified date for single post output.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @param string      $date_gmt GMT publication time.
+	 * @param string|null $date     Optional. Local publication time. Default null.
+	 * @return string|null ISO8601/RFC3339 formatted datetime.
+	 */
+	protected function prepare_date_response( $date_gmt, $date = null ) {
+		// Use the date if passed.
+		if ( isset( $date ) ) {
+			return mysql_to_rfc3339( $date );
+		}
+
+		// Return null if $date_gmt is empty/zeros.
+		if ( '0000-00-00 00:00:00' === $date_gmt ) {
+			return null;
+		}
+
+		// Return the formatted datetime.
+		return mysql_to_rfc3339( $date_gmt );
+	}
+
+	/**
+	 * Overwrites the default protected title format.
+	 *
+	 * By default, WordPress will show password protected posts with a title of
+	 * "Protected: %s". As the REST API communicates the protected status of a post
+	 * in a machine readable format, we remove the "Protected: " prefix.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @return string Protected title format.
+	 */
+	public function protected_title_format() {
+		return '%s';
+	}
+
+	/**
+	 * Retrieves the item schema, conforming to JSON Schema.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+
+		$schema = array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => 'search',
+			'type'       => 'object',
+			'properties' => array(
```

Comment:
I think we should drop: `author`, `content`, `excerpt`, `date`, `date_gmt`, `guid`, `modified`, `modified_gmt`, and `status` for a couple of reasons:

1. They aren't absolutely necessary for link search, which is the immediate problem we need to solve.
2. They'll significantly reduce complexity for the controller.

Reply:
I disagree as this is supposed to be merged into core. A REST controller in core should be available to cover common use-cases beyond Gutenberg. If this was a first test for Gutenberg internally, I'd be okay with using only the few fields needed for that very case, but for core, the typical properties expected should be available.

Furthermore, they don't really add complexity, but just increase the response. I'll implement the new `_fields` filtering so that those fields are not even generated when they're not needed. Gutenberg can that make a request like `wp/v2/search?search=mytext&_fields=id,title,type`.

Reply:
> I'd be okay with using only the few fields needed for that very case, but for core, the typical properties expected should be available.

Can you explain how you define "typical properties", and where, say, `modified_gmt` fits in this category?

Reply:
I'd define as typical properties those that exist on every post regardless of post type, i.e. things that are part of the `wp_posts` database table. In that regard I'd also like to align the format with that from the other post type-specific controllers.